open-models/serving/cloud_run_ollama_gemma2_rag_qa.ipynb

```diff
@@ -0,0 +1,1144 @@
+      },
+      "outputs": [],
+      "source": [
+        "loader = WebBaseLoader(\"https://cloud.google.com/run/docs/overview/what-is-cloud-run\")\n",
+        "docs = loader.load()\n",
+        "documents = CharacterTextSplitter(chunk_size=800, chunk_overlap=100).split_documents(\n",
+        "    docs\n",
+        ")\n",
+        "\n",
+        "vector = SKLearnVectorStore.from_documents(documents, embeddings)\n",
+        "retriever = vector.as_retriever()"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "rCH9n2UEzwMv"
+      },
+      "source": [
+        "### RAG Chain Definition\n",
+        "\n",
+        "We will define now our RAG Chain.\n",
+        "\n",
+        "The RAG chain works as follows:\n",
+        "\n",
+        "1. The user's query and conversation history are passed to the `query_rewrite_chain` to generate a rewritten query optimized for semantic search.\n",
+        "2. The rewritten query is used by the `retriever` to fetch relevant documents.\n",
+        "3. The retrieved documents are formatted into a single string.\n",
+        "4. The formatted documents, along with the original user messages, are passed to the LLM with instructions to generate an answer based on the provided context.\n",
+        "5. The LLM's response is parsed and returned as the final answer."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "x1nPO00s4r4o"
+      },
+      "outputs": [],
+      "source": [
+        "answer_generation_template = ChatPromptTemplate.from_messages(\n",
+        "    [\n",
+        "        (\n",
+        "            \"system\",\n",
+        "            \"You are an assistant for question answering-tasks. \"\n",
+        "            \"Use the following pieces of retrieved context to answer the question. If you don't know the answer, just say that you don't know. \"\n",
+        "            \"{context}\",\n",
+        "        ),\n",
+        "        MessagesPlaceholder(variable_name=\"messages\"),\n",
+        "    ]\n",
+        ")\n",
+        "query_rewrite_template = ChatPromptTemplate.from_messages(\n",
+        "    [\n",
+        "        (\n",
+        "            \"system\",\n",
+        "            \"Rewrite a query to a semantic search engine using the current conversation. \"\n",
+        "            \"Provide only the rewritten query as output.\",\n",
+        "        ),\n",
+        "        MessagesPlaceholder(variable_name=\"messages\"),\n",
+        "    ]\n",
+        ")"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "jm309i0r0Lwv"
+      },
+      "outputs": [],
+      "source": [
+        "query_rewrite_chain = query_rewrite_template | llm\n",
+        "\n",
+        "\n",
+        "def extract_query(messages):\n",
+        "    return query_rewrite_chain.invoke(messages).content\n",
```

Comment:
Nit: Can you add typing for the method parameters and return value?

Reply:
done, thanks!